### PR TITLE
feat: revamp PRQL codegen framework

### LINKED ISSUE #1420: PRQL code generation & formatting
> This is an onboarding issue, see instructions on [how to get started](https://github.com/PRQL/prql/blob/main/CONTRIBUTING.md#compiler).

For debugging purposes and for the `prqlc fmt` command, we have implemented a basic PRQL code generation from PL AST. It is done via `impl std::fmt::Display for X`, which is convenient, but lacks a few key features:
- indentation can only be done ad-hoc, when outputting nested pipelines or lists,
- we have no control over maximum line width,
- we have no control on tab size.

Instead, I suggest we follow implementation of [rustfmt](https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#a-quick-tour-of-rustfmt). I've started work on [codegen branch](https://github.com/PRQL/prql/compare/main...codegen).

- Start by implementing `WriteSource` for `pl::StmtKind`.
- You should be able to reuse majority of [`impl Display for StmtKind`](https://github.com/PRQL/prql/blob/9a024566c11f2fc3658f00d08f87af22d7e54f55/prql-compiler/src/ast/pl/stmt.rs#L94-L125) and adapt it to use string concatenation, similar to existing code in `src/codegen.rs`. Whit this I mean to change things like this:
  ```
  write!(f, "prql")?;
  if let Some(version) = &query.version {
      write!(f, " version:{}", version)?;
  }
  Ok(())
  ```
  ... into:
  ```
  let mut res = String::new();
  res += "prql";
  if let Some(version) = &query.version {
    res += &format!(" version:{}", version);
  }
  Some(res)
  ```

- When you use `write!(f, "{some_var}")` or `format!("{some_var}")`, Rust will call `std::fmt::Display` on `some_var`. We used this to do formatting and this issue is about migrating away from that to our own trait `WriteSource`. So when reusing the code from current impl, it's fine to use write! and format! for now, but the goal is to replace this with calls to `WriteSource::write()`.

- To test it, uncomment `r += &stmt.kind.write(opt.clone()).unwrap();` in `codegen.rs` and `cargo run fmt a_test_file.prql` should be working.

Here are a few test files:
```
prql    version:   "^0.4"    dialect:  sql.postgres

```

```
prql

from       my_table
select   [    some_column
]
```

```
prql    version:   "^0.4"    dialect:  sql.postgres

let my_table   =    (from base_table | select a)

from my_table | select [some_column]
```

The contents of pipelines are obviously not yet implemented and should output `<todo>` or something.

This is the gist of it. Similarly, we have adapt basically all `Display` impls in `ast::pl` to `WriteSource`. I suggest starting with `ExprKind`, which is already partially done and can easily be tested.

Because this is quite a lot of work, we can merge smaller commits implementing any small chunk of the codegen. 

This issue is a prerequisite for #130.

### PR DESCRIPTION
**Pull request by aljazerzen**
Closes #1420

### CODE REVIEW

File: prql-compiler/src/codegen.rs [hunk 1]
```diff
@@ -0,0 +1,590 @@
...
+
+impl std::fmt::Display for pl::Ty {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let opt = WriteOpt::new_width(u16::MAX);
+        f.write_str(&self.write(opt).unwrap())
+    }
+}
+
+pub trait WriteSource {
+    /// Converts self to it's source representation according to specified
```
Comment (max-sixty) [inline prql-compiler/src/codegen.rs:39]:
```suggestion
    /// Converts self to its source representation according to specified
```

File: prql-compiler/src/codegen.rs [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+    let needs_escape = s.is_empty()
+        || s.starts_with(forbidden_start)
+        || (s.len() > 1 && s.chars().skip(1).any(forbidden_subsequent));
+
+    if needs_escape {
+        format!("`{s}`")
+    } else {
+        s.to_string()
+    }
+}
```
Comment (max-sixty) [inline prql-compiler/src/codegen.rs:323]: GPT4 wrote this as a regex:

```suggestion
use regex::Regex;
use once_cell::sync::Lazy;

static IDENT_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[_$a-z][_a-z\d]*$").unwrap());

pub fn write_ident_part(s: &str) -> String {
    if IDENT_REGEX.is_match(s) {
        s.to_string()
    } else {
        format!("`{}`", s)
    }
}
```

I actually though I did this before, but probably it was for something different...

Comment (max-sixty) [reply prql-compiler/src/codegen.rs:323]: Should we also allow `*`, ref https://github.com/PRQL/prql/pull/2669/files#diff-107555a4ee015fdfa938922785f2d69c79432e2debb074e8270db0b05c9195c7R7 ?

**Comment by max-sixty** [comment]
Extremely impressive @aljazerzen !

File: prql-compiler/src/codegen.rs (hunk 2 above)
Comment (max-sixty) [reply prql-compiler/src/codegen.rs:323]: I remembered where I did this before — is this the same as https://github.com/prql/prql/blob/aa3ed6196596e38b3dfffad63d0b590e9eeed535/prql-compiler/src/sql/gen_expr.rs#L783 ? (fine to have it twice unless it's definitely going to remain the same)

Comment (aljazerzen) [reply prql-compiler/src/codegen.rs:323]: This is the problem with GPT: it looks likes it's correct, but it's not. The regex does not quote empty strings, but it should.

In anycase, this snippet was copied from the `gen_expr.rs` a while back and was living on `codegen` branch for last few months.

If you feel like it needs to be improved, go ahead. But add tests if you are using GPT.

Comment (max-sixty) [reply prql-compiler/src/codegen.rs:323]: Very good point!! 

Yeah, agree it has that frailty. I have found it helpful for these sorts of "chore" tasks; agree tests become more important...

Comment (max-sixty) [reply prql-compiler/src/codegen.rs:323]: Though in this case is it actually wrong? An empty string does get quoted by the regex.

Comment (aljazerzen) [reply prql-compiler/src/codegen.rs:323]: No, it's actually correct. An empty string does get quoted by the regex, which is correct. The logic has also been inverted.

But that's my point: it's more work validating the output as opposed to rewriting it. And things like this don't even need a rewrite or any more time spent on them.

Comment (max-sixty) [reply prql-compiler/src/codegen.rs:323]: Yes I agree re focus

Comment (max-sixty) [reply prql-compiler/src/codegen.rs:323]: (I had already put https://github.com/PRQL/prql/pull/2678 in — which I think does use the version that gets the `*` correct, but not 100% sure)

**Pull request merged**